[15.0][MIG] account_global_discount

kirca opened the pull request.

chienandalu committed:
[ADD][11.0] account_global_discount: New Module

chienandalu committed:
[FIX+IMP] account_global_discount: link line taxes to discount move line + Discounts in invoice report

[IMP] account_global_discount: Discounts in invoice report

Give discounts info in the invoice report.

[FIX] account_global_discount: link line taxes to discount move line

pedrobaeza committed:
[FIX] account_global_discount: More on global discount/taxes link

- We need to take into account invoice lines with multiple taxes, so the link should
  be m2m.
- Migration scripts for preserving the best information on existing installations.
- Tests for checking new conditions.
- Perform sanity check for not ending in an incompatible situation.
- Some refactor done on onchanges for avoiding duplicating operations.
- Adjust UI for not allowing to edit computed invoice global discounts.

pedrobaeza committed:
[MIG] acount_global_discount: Migration to 12.0

- Standard procedure
- Make test more resilient

pedrobaeza committed:
[IMP] account_global_discount: Deprecated use of currency compute

Replaced by more explicit `_convert` method.

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

pedrobaeza committed:
[FIX] account_global_discount: Fix singleton error in _compute_amount

For reducing diff, we rename the method and call it record per record.

Fixes #788

pedrocs-exo committed:
[FIX][12.0] account_global_discount: sale order templates compatibility

Commit:
[IMP] account_global_discount: black, isort, prettier

Commit:
[MIG] account_global_discount: Migration to 13.0

[UPD] Update account_global_discount.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/

chienandalu committed:
[IMP] account_global_discount: security group

Users will need a specific group to set global discounts so it's easier
to decide who's allowed to apply them.

[UPD] Update account_global_discount.pot

Commit:
[FIX] account_global_discount: wrong tax move line discounts on invoice post

OCA-git-bot committed:
account_global_discount 13.0.1.0.1

victoralmau committed:
[FIX] account_global_discount: Compute global discount with 0% taxes

Co-Authored-By: Pedro M. Baeza <pedro.baeza@tecnativa.com>

[UPD] README.rst

Commit:
Added translation using Weblate (Dutch)

Commit:
Translated using Weblate (Dutch)

Currently translated at 100.0% (35 of 35 strings)

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/nl_NL/

[UPD] Update account_global_discount.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/

anasuarez1 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (34 of 34 strings)

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/es/

victoralmau committed:
[FIX] account_global_discount: Fix tests

OCA-git-bot committed:
account_global_discount 13.0.1.0.2

chienandalu committed:
[FIX] account_global_discount: discounts on create

When we set the global discount on the create method we must ensure that
the move lines are properly recomputed

OCA-git-bot committed:
account_global_discount 13.0.1.1.0

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/

victoralmau committed:
[FIX] account_global_discount: Prevent error when add more taxes in multi-lines from invoices when global discount is empty

OCA-git-bot committed:
account_global_discount 13.0.1.1.1

pedrobaeza committed:
[FIX+REF] account_global_discount: Include taxes in global discount move lines

Previously, no taxes were populated as base taxes for global discount
move lines, which means that the tax reports were incorrect.

The global discount lines + move lines has been injected other way for
avoiding inconsistencies, and the rest of the features have been adapted
according this.

A migration script is provided as well for filling taxes in existing
global discount move lines.

[UPD] Update account_global_discount.pot

[UPD] README.rst

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-invoicing-13.0/account-invoicing-13.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-13-0/account-invoicing-13-0-account_global_discount/

pedrobaeza committed:
[FIX] account_global_discount: Call the proper method on create

For recomputing the global discounts, now we need to call another method.

TT30077

omar7r committed:
[MIG] account_global_discount: Migration to 14.0

omar7r committed:
[IMP] account_global_discount: securization for multi-company

omar7r committed:
[FIX] account_global_discount: Signed amount error in supplier invoices

omar7r committed:
[FIX] account_global_discount: computed fields must be stored because the other standard computed fields in same method are

omar7r committed:
[FIX] account_global_discount: Filtering global discount in on_change by invoice's company

omar7r committed:
[IMP] account_global_discount: Global discounts into invoice analysis report

omar7r committed:
[FIX] account_global_discount: Partner in global discount move line

omar7r committed:
[FIX] account_global_discount: When writing Odoo doesn't autocomplete amount_currency and it could fail with constraint

oca-travis committed:
[UPD] Update account_global_discount.pot

OCA-git-bot committed:
[UPD] README.rst

francesco-ooops committed:
Added translation using Weblate (Italian)

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (36 of 36 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_global_discount/it/

pedrobaeza committed:
[FIX] *: Change _recompute_tax_lines args according upstream

https://github.com/odoo/odoo/commit/c9d5f069739d897145f843ee93b7eefd1d39
has changed the number of arguments on the method `_recompute_tax_lines`,
so we need to adapt the overrides to such change.

OCA-git-bot committed:
account_global_discount 14.0.1.0.1

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (35 of 35 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_global_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_global_discount/pt/

kirca committed:
[IMP] account_global_discount: black, isort, prettier

kirca committed:
[MIG] account_global_discount: Migration to 15.0

omar7r commented on account_global_discount/views/report_account_invoice.xml:
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <template id="report_invoice_document" inherit_id="account.report_invoice_document">
+        <xpath expr="//div[@id='total']/div/table//tr[1]" position="before">
```
With this position in report, this information only shows when invoice already paid, in addtion, it shows after total amount. You should change it by: `//div[@id='total']/div/table/t[@t-set='tax_totals'] position="before"`

kirca replied:
thanks

victoralmau commented on account_global_discount/i18n/pt.po:
```diff
@@ -0,0 +1,221 @@
+#~ msgstr "<strong>Descontos Globais</strong><br/>"
+
+#~ msgid "Invoice Tax"
+#~ msgstr "Imposto de Fatura"
```
```suggestion
```

victoralmau commented on account_global_discount/i18n/nl_NL.po:
```diff
@@ -0,0 +1,218 @@
+#~ msgstr "Belasting"
+
+#~ msgid "Company related to this journal"
+#~ msgstr "Bedrijf gerelateerd aan journaal"
```
```suggestion
```

victoralmau commented on account_global_discount/i18n/es.po:
```diff
@@ -0,0 +1,223 @@
+
+#, fuzzy
+#~ msgid "Invoice Tax"
+#~ msgstr "Factura"
```
```suggestion
```

victoralmau commented on account_global_discount/models/account_move.py:
```diff
@@ -0,0 +1,400 @@
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    # HACK: Looks like UI doesn't behave well with Many2many fields and
+    # negative groups when the same field is shown. In this case, we want to
+    # show the readonly version to any not in the global discount group.
+    # TODO: Check if it's fixed in future versions
```
Have you checked if this is still happening in this version?

kirca replied:
It still seems to be an issue

victoralmau commented on account_global_discount/models/account_move.py:
```diff
@@ -0,0 +1,400 @@
+    )
+    def _compute_amount(self):
+        """Modify totals computation for including global discounts."""
+        __ = super()._compute_amount()
+        for record in self:
+            record._compute_amount_one()
+        return __
```
```suggestion
        res = super()._compute_amount()
        for record in self:
            record._compute_amount_one()
        return res
```

victoralmau commented on account_global_discount/models/account_move.py:
```diff
@@ -0,0 +1,400 @@
+
+    def _check_balanced(self):
+        """Add the check of proper taxes for global discounts."""
+        __ = super()._check_balanced()
```
```suggestion
        res = super()._check_balanced()
```

victoralmau commented on account_global_discount/models/account_move.py:
```diff
@@ -0,0 +1,400 @@
+                        )
+                else:
+                    taxes_keys[tuple(inv_line.tax_ids.ids)] = True
+        return __
```
```suggestion
        return res
```

kirca committed:
[FIX] account_global_discount: multi-currency

In a foreign currency invoice the amounts (amount_total,
amount_untaxed) are expressed in the `invoice` currency while the
accounting entry lines are expressed in the `company` currency.

The base_before_global_discounts has to be expressed in the invoice
currency as well in such invoices because the discounts need to be
computed in the same currency as the other amounts.

Also set the invoice currency_id on the tax line so that it doesn't
default to the company currency, and convert the discount journal entry
amount to the company currnecy.

alexeirivera87 committed:
[FIX] account_global_discount: avoid creation of undesired global discounts when the tax's computation is set to "Group of Taxes"

kirca committed:
[FIX] account_global_discount: show discounts in invoice report before total

+ update translations & small style changes

victoralmau reviewed (approved):
Code review OK

kirca commented:
Hi @amokaddem , why the additional commits? Last one breaks the pipeline

kirca commented:
Hello @omar7r , could you check if it is good now? Thanks

omar7r reviewed (approved):
LGTM, tested

rafaelbn commented:
/ocabot migration account_global_discount

rafaelbn reviewed (approved):
/ocabot merge patch
